third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -2887,21 +3008,61 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
     /// Adds conversions to the given arguments for the given resolved function entry. Currently
     /// the only supported conversion is from `&mut T` to `&T` and we treat with it in an ad-hoc
     /// manor.
-    fn add_conversions(&self, entry: &AnyFunEntry, instantiation: &[Type], args: &mut [Exp]) {
+    fn add_conversions(
+        &self,
+        entry: &AnyFunEntry,
+        instantiation: &[Type],
+        args: Vec<Exp>,
+    ) -> Vec<Exp> {
+        let mut new_args = vec![];
         let params = entry.get_signature().1;
         for (param_ty, exp) in params
             .iter()
             .map(|Parameter(_, ty, _)| ty.instantiate(instantiation))
-            .zip(args.iter_mut())
+            .zip(args)
         {
             let exp_id = exp.node_id();
             let exp_ty = self.env().get_node_type(exp_id);
-            if exp_ty.is_mutable_reference() && param_ty.is_immutable_reference() {
-                // Insert Freeze operation
-                let new_id =
-                    self.new_node_id_with_type_loc(&param_ty, &self.env().get_node_loc(exp_id));
-                *exp = ExpData::Call(new_id, Operation::Freeze, vec![exp.clone()]).into_exp();
-            }
+            new_args.push(self.try_freeze(&param_ty, &exp_ty, exp));
+        }
+        new_args
+    }
+
+    /// Inserts the freeze operation when `expected_ty` is immutable ref and ty is mutable ref
+    fn try_freeze(&self, expected_ty: &Type, ty: &Type, exp: Exp) -> Exp {
+        if expected_ty.is_immutable_reference() && ty.is_mutable_reference() {
+            let exp_id = exp.node_id();
+            let new_id =
+                self.new_node_id_with_type_loc(expected_ty, &self.env().get_node_loc(exp_id));
+            ExpData::Call(new_id, Operation::Freeze, vec![exp]).into_exp()
+        } else {
+            exp
+        }
+    }
+
+    /// Inserts the freeze operation when `exp` is a tuple expression
+    fn freeze_tuple_exp(
+        &self,
+        lhs_tys: &Vec<Type>,
+        rhs_tys: &Vec<Type>,
+        exp: Exp,
+        loc: &Loc,
+    ) -> Exp {
+        if lhs_tys.len() != rhs_tys.len() || lhs_tys.eq(rhs_tys) {
+            return exp;
+        }
+        if let ExpData::Call(_, Operation::Tuple, rhs_vec) = exp.as_ref() {
```

Comment:
Note that we're making a new vector and `Call(...Tuple..)` expression below for *every* tuple assignment, even if it doesn't have reference types.  While `Exp` is interned, its constructors are not memoized, so we are generating a lot of duplicated structures (which keep using memory, unfortunately, due to `Exp` interning).

Maybe we need a function `need_freeze(lh_ty, rh_ty) -> bool` that we can use to first test whether we even need the new expression:
```
    let need_freeze = lhs_tys.iter().zip(rhs_tys.iter()).any(|(lh_ty, rh_ty)| self.need_freeze(lh_ty, rh_ty)));
    if let (true, ExpData::Call(_, Operation::Tuple, rhs_vec)) = (need_freeze, exp.as_ref()) {
        ...
    }
```